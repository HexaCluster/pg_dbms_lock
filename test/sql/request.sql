-----
-- Test exclusive lock request+release at session level
-----
DO $$
DECLARE
    lock_res int;
    DBMS_LOCK_X_MODE int := 6;
BEGIN
    lock_res := DBMS_LOCK.REQUEST( 123, DBMS_LOCK_X_MODE, 300, FALSE );
    IF ( lock_res <> 0 ) THEN
	RAISE EXCEPTION 'DBMS_LOCK.REQUEST() FAIL: %', lock_res;
    END IF;
END;
$$;

SELECT objid, mode FROM pg_locks WHERE objid IS NOT NULL AND locktype = 'advisory' ORDER BY objid;

DO $$
DECLARE
    lock_res int;
BEGIN
    -- release lock
    lock_res := DBMS_LOCK.RELEASE( 123 );
    IF ( lock_res <> 0 ) THEN
        RAISE EXCEPTION 'DBMS_LOCK.RELEASE() FAIL: %', lock_res;
    END IF;
END;
$$;

SELECT objid, mode FROM pg_locks WHERE objid IS NOT NULL AND locktype = 'advisory' ORDER BY objid;

----
-- Test shared lock request+release at session level
----
DO $$
DECLARE
    lock_res int;
    DBMS_LOCK_S_MODE int := 4;
BEGIN
    lock_res := DBMS_LOCK.REQUEST( 321, DBMS_LOCK_S_MODE, 300, FALSE );
    IF ( lock_res <> 0 ) THEN
	RAISE EXCEPTION 'DBMS_LOCK.REQUEST() FAIL: %', lock_res;
    END IF;
END;
$$;

SELECT objid, mode FROM pg_locks WHERE objid IS NOT NULL AND locktype = 'advisory' ORDER BY objid;

DO $$
DECLARE
    lock_res int;
BEGIN
    -- release lock
    lock_res := DBMS_LOCK.RELEASE( 321 );
    IF ( lock_res <> 0 ) THEN
        RAISE EXCEPTION 'DBMS_LOCK.RELEASE() FAIL: %', lock_res;
    END IF;
END;
$$;

SELECT objid, mode FROM pg_locks WHERE objid IS NOT NULL AND locktype = 'advisory' ORDER BY objid;

-----
-- Test xact exclusive lock request at transaction level
-----
BEGIN;

DO $$
DECLARE
    lock_res int;
    DBMS_LOCK_X_MODE int := 6;
BEGIN
    lock_res := DBMS_LOCK.REQUEST( 321, DBMS_LOCK_X_MODE, 300, TRUE );
    IF ( lock_res <> 0 ) THEN
	RAISE EXCEPTION 'DBMS_LOCK.REQUEST() FAIL: %', lock_res;
    END IF;
END;
$$;

SELECT objid, mode FROM pg_locks WHERE objid IS NOT NULL AND locktype = 'advisory' ORDER BY objid;

COMMIT;

SELECT objid, mode FROM pg_locks WHERE objid IS NOT NULL AND locktype = 'advisory' ORDER BY objid;

-----
-- Test xact shared lock request at transaction level
-----
BEGIN;

DO $$
DECLARE
    lock_res int;
    DBMS_LOCK_S_MODE int := 4;
BEGIN
    lock_res := DBMS_LOCK.REQUEST( 321, DBMS_LOCK_S_MODE, 300, TRUE );
    IF ( lock_res <> 0 ) THEN
	RAISE EXCEPTION 'DBMS_LOCK.REQUEST() FAIL: %', lock_res;
    END IF;
END;
$$;

SELECT objid, mode FROM pg_locks WHERE objid IS NOT NULL AND locktype = 'advisory' ORDER BY objid;

COMMIT;

SELECT objid, mode FROM pg_locks WHERE objid IS NOT NULL AND locktype = 'advisory' ORDER BY objid;
