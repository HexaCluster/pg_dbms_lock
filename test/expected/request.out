DO
 objid |     mode      
-------+---------------
   123 | ExclusiveLock
(1 row)

DO
 objid | mode 
-------+------
(0 rows)

DO
 objid |   mode    
-------+-----------
   321 | ShareLock
(1 row)

DO
 objid | mode 
-------+------
(0 rows)

BEGIN
DO
 objid |     mode      
-------+---------------
   321 | ExclusiveLock
(1 row)

COMMIT
 objid | mode 
-------+------
(0 rows)

BEGIN
DO
 objid |   mode    
-------+-----------
   321 | ShareLock
(1 row)

COMMIT
 objid | mode 
-------+------
(0 rows)

